File: Cliente.java
package RockPaperScissorsLizardSpock;
import java.net.*;
import java.io.*;
import java.text.DecimalFormat;
import java.util.Random;
import java.util.Scanner;
import java.lang.Thread;
import java.util.Vector;
import java.util.concurrent.ThreadLocalRandom;

class Dados {
    public String my_jogada;
    public String op_jogada;
    public int result;
    public Dados(String my,String op,int res) {
        my_jogada = my;
        op_jogada = op;
        result = res;
    }
}

public class Cliente {
    int sleep_time = 100;
    int limite_games = 15;
    Vector<Dados> historico;
    String HOST;
    int PORT;
    String clientID;
    Boolean active = true;
    int games_played = 0;
    int status = 0;
    int player = 0;
    String my_last_play = "";
    String op_last_play = "";
    int res_last_play = 1;
    DecimalFormat df;
    public Cliente(String host, int port, String id) {
        df = new DecimalFormat();
        df.setMaximumFractionDigits(2);

        historico = new Vector<>();
        this.HOST = host;
        this.PORT = port;
        this.clientID = id;
        System.out.println(String.format("Cliente %s ativado", id));
        Scanner scan = new Scanner(System.in);
        while (games_played < limite_games) {
            try {
                Thread.sleep(sleep_time);
                String line = obterMensagem();
                String mensagemFinal = this.clientID + '-' + line;
                Socket socket = new Socket(host, port);
                PrintStream out = new PrintStream( socket.getOutputStream() );
                BufferedReader in = new BufferedReader( new InputStreamReader( socket.getInputStream() ) );
                out.print(mensagemFinal);
                String response = in.readLine();
                lidarComResposta(response);
                in.close();
                out.close();
                socket.close();
            }
            catch(UnknownHostException u) {
                System.out.print(".");
            }
            catch(IOException i) {
                System.out.print(".");
            }
            catch (InterruptedException i) {
                System.out.print(".");
            }
        }
        int vitorias = 0;
        int derrotas = 0;
        int empates = 0;
        int i = 0;
        System.out.println("\nFim do Jogo! Apresentando as partidas realizadas:");
        for (Dados dado: historico) {
            i+=1;
            String resultado = "Empate! (Nao conta como rodada)";
            if (dado.result == 2) {
                vitorias += 1;
                resultado = "Vitoria!";
            }
            else if (dado.result == 3) {
                derrotas += 1;
                resultado = "Derrota!";
            }
            else {
                empates += 1;
            }
            System.out.print("\nPartida "+ i + " - " + dado.my_jogada + " x " + dado.op_jogada + " - " + resultado);
        }
        if (vitorias - derrotas > 0) {
            System.out.println("\nVitorioso!");
        }
        else if (vitorias - derrotas < 0) {
            System.out.println("\nDerrotado!");
        }
        else {
            System.out.println("\nEmpatados!");
        }
        System.out.println("\nVitorias: " + df.format(100*((float)vitorias / (vitorias+empates+derrotas)))+"%");
        System.out.println("Derrotas: " + df.format(100*((float)derrotas / (vitorias+empates+derrotas)))+"%");
        System.out.println("Empates: " + df.format(100*((float)empates / (vitorias+empates+derrotas)))+"%");
    }
    public String getRandomElement() {
        String[] arr = {"Rock", "Paper", "Scissors", "Lizard", "Spock"};
        return arr[ThreadLocalRandom.current().nextInt(arr.length)];
    }
    // Heuristica, se vencer, pega o elemento que do opponente
    // Se perder, pega algo que no foi pego pelos dois
    public String heuristica() {
        if (my_last_play.equals("") || res_last_play == 1) {
            return getRandomElement();
        }
        else if (res_last_play == 2) {
            return op_last_play;
        }
        else {
            String ret = my_last_play;
            while (ret.equals(my_last_play) || ret.equals(op_last_play)){
                ret = getRandomElement();
            }
            return ret;
        }
    }

    public void printResumo() {
        int vitorias = 0;
        int derrotas = 0;
        int empates = 0;
        int i = 0;
        for (Dados dado: historico) {
            i+=1;
            if (dado.result == 2) {
                vitorias += 1;
            }
            else if (dado.result == 3) {
                derrotas += 1;
            }
            else {
                empates += 1;
            }
        }
        int rodada = vitorias + derrotas;
        int total = vitorias+empates+derrotas;
        System.out.print("\nRodada " + rodada + " - ");
        System.out.print("Vitorias: " + df.format(100*((float)vitorias / (total)))+"% ");
        System.out.print("Derrotas: " + df.format(100*(float)derrotas / (total))+"% ");
        System.out.print("Empates: " + df.format(100*(float)empates / (total))+"%");
    }
    public String obterMensagem() {
        if (status == 0) {
            return "connect";
        }
        else if (status == 1) {
            return heuristica();
        }
        else {
            return "get";
        }
    }
    public void lidarComResposta(String resposta) {
        if (status == 0) {
            if (resposta.contains("connected as player")){
                status = 1;
            }
        }
        else if (status == 1) {
            if (resposta.contains("to get result, send message 'get'")) {
                status = 2;
                if (resposta.contains("1")) {
                    player = 1;
                }
                else {
                    player = 2;
                }
            }
        }
        else {
            int resp = extractFromString(resposta);
            if (resp != 1) {
                printResumo();
                games_played += 1;
            }
            status = 0;
        }
    }

    public int extractFromString(String str) {
        String[] str_parantheses = str.substring(str.indexOf("(")+1, str.indexOf(")")).split(" x ");
        String my_play = str_parantheses[0];
        String op_play = str_parantheses[1];
        if (player == 2) {
            String swap = my_play;
            my_play = op_play;
            op_play = swap;
        }
        int result = 1;
        if (str.contains("you won")) {
            result = 2;
        }
        if (str.contains("you lost")) {
            result = 3;
        }
        saveResult(my_play, op_play, result);
        return result;
    }

    public void saveResult(String my_play, String op_play, int result) {
        my_last_play = my_play;
        op_last_play = op_play;
        res_last_play = result;
        Dados d = new Dados(my_play,op_play,result);
        historico.add(d);
    }
    public static String getRandom(int max) {
        return String.valueOf((int)(Math.random() * max + 1));
    }

    public static void main(String args[])
    {
        Cliente client = new Cliente("172.15.3.50", 12000, "java_gamer".concat(getRandom(1000)));
    }
}

File: cliente.py
from socket import *
import threading
import random
from time import sleep
import sqlite3

SLEEP_TIME = 0.1
LIMITE_JOGOS = 15

def getRandomNumber():
  return random.randint(1, 1000)
  
class handlerClient(threading.Thread):
  def __init__(self, PORT, HOST, ID):
    super(handlerClient, self).__init__()
    self.active = True
    self.HOST = HOST
    self.PORT = PORT
    if ('-' in ID):
      self.ID = ID.replace('-', '')
    self.clientID = str(ID)
    self.numero_jogos = 0
    self.status = 0
    self.player = 0
  
  def extractFromString(self, str):
    str_paratheses = str[str.find("(")+1:str.find(")")].split(" x ")
    my_play = str_paratheses[0]
    op_play = str_paratheses[1]
    if (self.player == 2):
      swap = my_play
      my_play = op_play
      op_play = swap
    result = 1 # 1 = empate, 2 = ganhei, 3 = perdi
    if ("you won" in str):
      result = 2
    if ("you lost" in str):
      result = 3
    self.saveResult(my_play, op_play, result)
    return result
  
  def createBanco(self):
    try:
        sqliteConnection = sqlite3.connect('banco_de_dados.db')
        c = sqliteConnection.cursor() 
        c.execute("""create table if not exists cliente_dados (
                     rodada integer primary key autoincrement ,
                     my_play text,
                     op_play text,
                     resultado integer)""")
        sqliteConnection.commit()
        c.close()
    except sqlite3.Error as error:
        print("Error while connecting to sqlite", error)
    finally:
        if sqliteConnection:
            sqliteConnection.close()
  
  def saveResult(self, my_play, op_play, resultado):
    self.createBanco()
    try:
        sqliteConnection = sqlite3.connect('banco_de_dados.db')
        c = sqliteConnection.cursor() 
        data = [(my_play, op_play, resultado)]
        insertstring = "INSERT INTO cliente_dados (my_play, op_play, resultado) VALUES (?, ?, ?)"
        c.executemany(insertstring, data)
        sqliteConnection.commit()
        c.close()
    except sqlite3.Error as error:
        print("Error while inserting to sqlite", error)
    finally:
        if sqliteConnection:
            sqliteConnection.close()
  
  # Estrategia, ver o mais frequente das ultimas, jogar o oposto, so se rodadas > 5
  def heuristica(self):
    self.createBanco()
    text = ''
    plays = [0,0,0,0,0]
    amount = 0
    try:
      sqliteConnection = sqlite3.connect('banco_de_dados.db')
      c = sqliteConnection.cursor() 
      c.execute("select * from cliente_dados")
      lista = c.fetchall()
      for (item) in lista:
        amount += 1
        op_play = item[2]
        if (op_play == 'Rock'):
          plays[0] += 1
        if (op_play == 'Paper'):
          plays[1] += 1
        if (op_play == 'Scissors'):
          plays[2] += 1
        if (op_play == 'Lizard'):
          plays[3] += 1
        if (op_play == 'Spock'):
          plays[4] += 1
        text = f"{text} || {item[0]} - {item[1]} vs {item[2]} - {item[3]}!"
      c.close()
    except sqlite3.Error as error:
      print("Error while inserting to sqlite", error)
    finally:
      if sqliteConnection:
          sqliteConnection.close()
    if (amount > 5):
      indice = plays.index(max(plays))
      if (indice == 0): # case rock
        possibilities = ['Paper', 'Spock']
      if (indice == 1): # case paper
        possibilities = ['Scissors', 'Lizard']
      if (indice == 2): # case scissors
        possibilities = ['Rock', 'Spock']
      if (indice == 3): # case lizard
        possibilities = ['Rock', 'Scissors']
      if (indice == 4): # case spock
        possibilities = ['Paper', 'Lizard']
      random_number = random.randint(0, 1)
      return possibilities[random_number]
    else:
      random_number = random.randint(1, 5)
      if (random_number == 1):
        return 'Paper'
      if (random_number == 2):
        return 'Scissors'
      if (random_number == 3):
        return 'Lizard'
      if (random_number == 4):
        return 'Spock'
      if (random_number == 5):
        return 'Rock'
  
  def sendMessage(self, s):
    s.connect((self.HOST, self.PORT))
    if self.status == 0:
      data = "connect"
      s.sendall(f"{self.clientID}-{data}".encode())
      resposta = s.recv(1024).decode().replace('\n', '')
      if ("connected as player" in resposta):
        self.status = 1
        return
    if self.status == 1:
      data = self.heuristica()
      s.sendall(f"{self.clientID}-{data}".encode())
      resposta = s.recv(1024).decode().replace('\n', '')
      if ("to get result, send message 'get'" in resposta):
        self.status = 2
        if ('1' in resposta):
          self.player = 1
        else:
          self.player = 2
        return
    if self.status == 2:
      data = 'get'
      s.sendall(f"{self.clientID}-{data}".encode())
      resposta = s.recv(1024).decode().replace('\n', '')
      resp = self.extractFromString(resposta)
      if (resp != 1):
        self.printResumo()
        self.numero_jogos += 1
      self.status = 0
  
  def printResults(self):
    self.createBanco()
    try:
      win = 0
      lose = 0
      draw = 0
      sqliteConnection = sqlite3.connect('banco_de_dados.db')
      c = sqliteConnection.cursor() 
      c.execute("select * from cliente_dados")
      lista = c.fetchall()
      i = 0
      print("\nFim do Jogo! Apresentando as partidas realizadas:")
      for (item) in lista:
        i+=1
        if (item[3] == 2):
          win += 1
        elif (item[3] == 3):
          lose += 1
        else:
          draw += 1
        print(f"\nPartida {i} - {item[1]} x {item[2]} - Result {'Empate' if item[3] == 1 else ('Vitoria' if item[3] == 2 else 'Derrota')}", end="")
      c.close()
    except sqlite3.Error as error:
      print("Error while inserting to sqlite", error)
    finally:
      if sqliteConnection:
          sqliteConnection.close()
    points = win - lose
    total = win + lose + draw
    print("\n" + f"{'Vitorioso!' if points > 0 else 'Derrotado!' if points < 0 else 'Empatado!'}")
    print(f"\nVitorias: {round(100*win/(total),2)}% \nDerrotas: {round(100*lose/total, 2)}% \nEmpates: {round(100*draw/total, 2)}%")
    
  def clearBanco(self):
    self.createBanco()
    try:
      sqliteConnection = sqlite3.connect('banco_de_dados.db')
      c = sqliteConnection.cursor() 
      c.execute("delete from cliente_dados")
      sqliteConnection.commit()
    except sqlite3.Error as error:
      print("Error while inserting to sqlite", error)
    finally:
      if sqliteConnection:
          sqliteConnection.close()
  
  def printResumo(self):
    self.createBanco()
    try:
      win = 0
      lose = 0
      draw = 0
      sqliteConnection = sqlite3.connect('banco_de_dados.db')
      c = sqliteConnection.cursor() 
      c.execute("select * from cliente_dados")
      lista = c.fetchall()
      i = 0
      for (item) in lista:
        i+=1
        if (item[3] == 2):
          win += 1
        elif (item[3] == 3):
          lose += 1
        else:
          draw += 1
      c.close()
    except sqlite3.Error as error:
      print("Error while inserting to sqlite", error)
    finally:
      if sqliteConnection:
          sqliteConnection.close()
    total = win + lose + draw
    print(f"\nRodada {win + lose} - Vitorias: {round(100*win/(total),2)}% | Derrotas: {round(100*lose/total,2)}% | Empates: {round(100*draw/total,2)}%", end="")
  
  def run(self):
    self.clearBanco()
    print(f"Cliente {self.clientID} ativado")
    while self.numero_jogos < LIMITE_JOGOS:
      try:
        sleep(SLEEP_TIME)
        with socket(AF_INET, SOCK_STREAM) as s:
          self.sendMessage(s)
          
      except Exception as e:
        if (self.numero_jogos < LIMITE_JOGOS):
          print(".", end="")
          sleep(1)
    self.printResults()
if __name__ == "__main__":
  cliente = handlerClient(PORT=12000, HOST="172.15.3.50", ID=f"python_gamer{str(getRandomNumber())}")
  cliente.start()
  cliente.join()
  print("Ending!")
  
